### PR TITLE
[Metricbeat] [Statsd] Add support for Graphite series 1.1.0+ tags

### LINKED ISSUE #39473: [Metricbeat] Statsd Module does not support Graphite series 1.1.0+
When using the Metricbeat statsd module to ingest Envoy statistics I am unable to use the [Graphite Statsd extension](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/stat_sinks/graphite_statsd/v3/graphite_statsd.proto#extension-envoy-stat-sinks-graphite-statsd). 

Using the [Statsd Sink ](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/metrics/v3/stats.proto#extension-envoy-stat-sinks-statsd) does work however but for more efficeint parsing it would be best to use Graphite.

Here is an example output for the Graphiite Statsd extension

```
envoy.listener.ssl.ocsp_staple_omitted;envoy.listener_address=0.0.0.0_9001:0|c
```

The statsd module is splitting on `,` while Graphite supports `;` seen [here](https://graphite.readthedocs.io/en/latest/tags.html#carbon)

The [`parseSingle`](https://github.com/elastic/beats/blob/13aed9f010453e05783ca3b318651ec2949030c4/x-pack/metricbeat/module/statsd/server/data.go#L76) function is only splitting on `,` as mentioned before and would be great if we could also support `;`

Please include configurations and logs if available.

For confirmed bugs, please report:
- Version: 
  - Envoy: 1.20.1
  - Metricbeat: 8.13.2
- Operating System: MacOS 14.4.1
- Discuss Forum URL: N/A
- Steps to Reproduce: 

1. Add the graphite statsd config to envoy

```yaml
stats_sinks:
  - name: graphite_statsd
    typed_config:
      "@type": type.googleapis.com/envoy.extensions.stat_sinks.graphite_statsd.v3.GraphiteStatsdSink
      address:
        socket_address:
          address: 127.0.0.1
          port_value: 8125
```

2. Enable statsd metricbeat module

```yaml
- module: statsd
  host: "localhost"
  metricsets: ["server"]
  port: "8125"
  enabled: true
```

3. Run envoy and metricbeat
4. Result is the following

ES event

### PR DESCRIPTION
**Pull request by tehbooom**
## Proposed commit message

Added conditional to check if  `statsd` metric contains `,` or `;` and split accordingly

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Closes #39473 

## Screenshots

Screenshot 2024-05-17 at 11 14 46 AM

## Logs

```json
{
  "_index": ".ds-metricbeat-8.15.0-2024.05.17-000001",
  "_id": "iNaPho8Bf8HHEaVjdnAn",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2024-05-17T12:38:09.825Z",
    "service": {
      "type": "statsd"
    },
    "statsd": {
      "envoy_listener_worker_downstream_cx_active": {
        "value": 0
      }
    },
    "labels": {
      "envoy.worker_id": "10",
      "envoy.listener_address": "0.0.0.0_9001"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "host": {
      "name": "mbp"
    },
    "agent": {
      "name": "mbp",
      "type": "metricbeat",
      "version": "8.15.0",
      "ephemeral_id": "41c29b63-8a18-4387-9879-a1b212dbcf8a",
      "id": "a934cbff-9d5b-4b73-84dd-363af4de9d23"
    },
    "metricset": {
      "name": "server"
    },
    "event": {
      "dataset": "statsd",
      "module": "statsd"
    }
  }
}
```

### CODE REVIEW

**Comment by ritalwar** [comment]
Can you also please update the changelog?

File: CHANGELOG.next.asciidoc
Comment (ritalwar) [inline CHANGELOG.next.asciidoc:51]:
```suggestion
- Added a conditional check to see if the statsd metric contains ',' or ';' and split accordingly. {pull}39619[39619]
```

**Review by ritalwar — COMMENTED** [review]
LGTM!

**Comment by ritalwar** [comment]
Hi @aliabbas-elastic , Can we verify these changes? I've opened an [issue](https://github.com/elastic/beats/issues/39681) for it here.
cc: @lalit-satapathy

**Comment by aliabbas-elastic** [comment]
> Hi @aliabbas-elastic , Can we verify these changes? I've opened an https://github.com/elastic/beats/issues/39681 for it here.

@ritalwar Ok I'll look into it

File: x-pack/metricbeat/module/statsd/server/data.go
Comment (shmsr) [inline x-pack/metricbeat/module/statsd/server/data.go:89]:
```suggestion
```

File: x-pack/metricbeat/module/statsd/server/data.go
Comment (shmsr) [inline x-pack/metricbeat/module/statsd/server/data.go:52]: We can reduce the code repetition like this:

```go
func splitTags(rawTags, kvSep []byte) map[string]string {
    tags := map[string]string{}
    var tagSplit [][]byte

    if bytes.Contains(rawTags, []byte(",")) {
        tagSplit = bytes.Split(rawTags, []byte(","))
    } else { // or use else if??
        tagSplit = bytes.Split(rawTags, []byte(";"))
    }

    for _, kv := range tagSplit {
        kvSplit := bytes.SplitN(kv, kvSep, 2)
        if len(kvSplit) != 2 {
            logger.Warn("could not parse tags")
            continue
        }
        tags[string(kvSplit[0])] = string(kvSplit[1])
    }

    return tags
}
``` 


I am just thinking about something. Is it possible to have an escaped separator in the tag's key/value? If yes, then should we handle? Maybe not as part of this PR. Code change is simple.

In the loop, we just have to do:
```
kv = bytes.ReplaceAll(kv, []byte("\\,"), []byte(","))
kv = bytes.ReplaceAll(kv, []byte("\\;"), []byte(";"))
kv = bytes.ReplaceAll(kv, []byte("\\:"), []byte(":"))
```

File: x-pack/metricbeat/module/statsd/server/data.go
Comment (shmsr) [inline x-pack/metricbeat/module/statsd/server/data.go:92]: Add a comment here as well.

File: CHANGELOG.next.asciidoc
Comment (shmsr) [inline CHANGELOG.next.asciidoc:51]:
```suggestion
- Add support of Graphite series 1.1.0+ tagging extension for statsd module. {pull}39619[39619]
```

Comment (shmsr) [reply CHANGELOG.next.asciidoc:51]: Also, I think we should update README. See an example: https://github.com/prometheus/statsd_exporter?tab=readme-ov-file#tagging-extensions

@ritalwar @tehbooom What do you think? Else, one has to go to code to figure it out to see what extensions are supported.

File: x-pack/metricbeat/module/statsd/server/data.go
Comment (tehbooom) [reply x-pack/metricbeat/module/statsd/server/data.go:52]: The only example I can find or test for myself is if the metric name has `:` as seen [here](https://github.com/statsd/statsd/issues/585)

File: CHANGELOG.next.asciidoc
Comment (tehbooom) [reply CHANGELOG.next.asciidoc:51]: @ritalwar @shmsr How do we feel about adding this to the README?

```asciidoc
[float]
=== Suported tag extensions

The `statsd` module supports the following tags:

https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/?tab=metrics#the-dogstatsd-protocol[DogStatsD]

`<metric name>:<value>|<type>|@samplerate|#<k>:<v>,<k>:<v>`

https://github.com/influxdata/telegraf/blob/master/plugins/inputs/statsd/README.md#influx-statsd[InfluxDB]

`<metric name>,<k>=<v>,<k>=<v>:<value>|<type>|@samplerate`

https://graphite.readthedocs.io/en/latest/tags.html#graphite-tag-support[Graphite 1.1.x]

`<metric name>;<k>=<v>;<k>=<v>:<value>|<type>|@samplerate`
```

Comment (shmsr) [reply CHANGELOG.next.asciidoc:51]: Perhaps the heading could be like:

- "DogStatsD-style tags"
- "InfluxDB-style tags"

and so on.

File: x-pack/metricbeat/module/statsd/server/data.go
Comment (shmsr) [reply x-pack/metricbeat/module/statsd/server/data.go:52]: Yeah, perhaps we should do this enhancement in a later PR. It immediately hit me and thought the logic should be there.

Anyway, for now, you can avoid repetition in code by using the snippet I gave above.

File: CHANGELOG.next.asciidoc
Comment (ritalwar) [reply CHANGELOG.next.asciidoc]: I'm fine with just listing the names, but we can add something like:

Supported Tag Extensions

Examples of tag styles supported by the `statsd` module:

- DogStatsD 
`<metric name>:<value>|<type>|@samplerate|#<k>:<v>,<k>:<v>`
- InfluxDB
`<metric name>,<k>=<v>,<k>=<v>:<value>|<type>|@samplerate`
- Graphite 1.1.x
`<metric name>;<k>=<v>;<k>=<v>:<value>|<type>|@samplerate`

File: x-pack/metricbeat/module/statsd/server/data.go
Comment (shmsr) [inline x-pack/metricbeat/module/statsd/server/data.go:86]:
```suggestion
	// Metric tags could be separated by `,` or `;`, we split here based on the separator.
```

**Review by shmsr — APPROVED** [review]
LGTM. Let's wait for @ritalwar to approval as well and then we can merge.

**Review by ritalwar — APPROVED** [review]

**Comment by ritalwar** [comment]
@aliabbas-elastic, could you please do one round of testing as well so we can merge this?

File: metricbeat/docs/modules/statsd.asciidoc
Comment (aliabbas-elastic) [inline metricbeat/docs/modules/statsd.asciidoc:32]:
```suggestion
=== Supported tag extensions
```
Minor nit

**Review by aliabbas-elastic — APPROVED** [review]
Approving this PR but let me run one round of sanity tests for compatible metric formats. I'll update [here](https://github.com/elastic/beats/issues/39681) soon

**Comment by shmsr** [comment]
Yes, let's wait for @aliabbas-elastic before merging

**Pull request merged**
